base/abstractarray.jl

```diff
@@ -173,6 +173,10 @@ For dictionary types, this will be a `Pair{KeyType,ValType}`. The definition
 instead of types. However the form that accepts a type argument should be defined for new
 types.
 
+For consistency it is generaly advised that for iterable objects `eltype` satisfies 
+`eltype(typeof(itr)) == typeof(first(itr))` and if `itr` implements `Base.getindex`, then also
+`eltype(typeof(itr)) == typeof(itr[a_valid_index])` holds.
```

Comment:
I don't think that is correct. The latter just has to be a subtype of the former.

Reply:
I learned about these assumptions when broadcasting on a struct of mine got broken in julia-1.6. Could you point me to any reference for `typeof(first(itr)) <: eltype(typeof(itr))`?

Reply:
The later is not required, or expected, to hold true. The `eltype` is for iteration, not indexing.

Reply:
What he means is that eltype may be any supertype, for example, it may be `Any`. What is expected to hold true is:
```julia
all(isa(eltype(itr)), itr)
```

Reply:
Yes subtyping is enough. But why should `eltype` apply to iteration but not indexing @vtjnash? I don't really care, but that sounds natural.

Reply:
I updated the docstrings, according to @simeonschaub and @vtjnash suggestions. How about `getindex`? should it remain as recommendation or formal requirement?

(I know that I may be opening Pandora's box ;)

Reply:
It's **definitely** a formal requirement — for both iteration and (scalar) indexing — if you're a subtype of `AbstractArray` (or `Number`).

> The eltype is for iteration, not indexing.

???

Reply:
Right, for some types, those will be the same, but I think that's a derived property, not a requirement:

> For types, such as `AbstractArray`, where the iteration space can also be indexed directly, this implies it is also the type of the result of `getindex` for one of those indexes.

Otherwise, this definition becomes mostly tautological ("scalar indexes are those keys that return elements whose types are eltype / the values returned by indexing with a scalar index have type eltype").

Else, there are numerous counter-examples where the requirement doesn't hold, even for array (such as indexing with a BitArray or Vector)

Reply:
> Else, there are numerous counter-examples where the requirement doesn't hold, even for array (such as indexing with a BitArray or Vector)

ok,I thought that the docstring could say
```
`typeof(itr[idx]) <: eltype(typeof(itr)) ` holds for each `idx` in `eachindex(itr)`.
```
but indexable types do not necessarily implement `eachindex`...

Reply:
This is largely already documented in https://docs.julialang.org/en/v1/manual/interfaces/ — but again, what's required is dependent upon the abstract super type.

Reply:
https://github.com/JuliaLang/julia/pull/39185 tries to document a case that was broken by 1.6 and does follow everything(?) what is in written in https://docs.julialang.org/en/v1/manual/interfaces/.
I'll wait for the conclusion of discussions in #39185 on what to write here exactly.

Reply:
Sorry, I didn't mean to imply that this _shouldn't_ be documented here — just that the requirements depend upon the interface and it's hard to document exactly what this means without reference to this.

Reply:
@mbauman no offence was taken ;)
Are there any discussions here (github) or elsewhere on `Number` interface? It seems that this issue should belong there.